group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/RangeAssignor.java

```diff
@@ -0,0 +1,240 @@
+
+        membersPerTopic.forEach((topicId, membersForTopic) -> {
+            // For each topic we have a temporary list of members stored in potentiallyUnfilledMembers.
+            // The list is populated with members that satisfy one of the two conditions:
+            // 1) Members that already have the minimum required number of partitions i.e. total partitions divided by total subscribed members
+            //    BUT they could be assigned an extra partition later on. Extra partitions exist if total partitions % number of members is greater than 0.
+            //    In this case we add the member to the unfilled members map iff an extra partition needs to be assigned to it.
+            // 2) Members that don't have the minimum required partitions, so irrespective of whether they get an extra partition or not they get added to the unfilled map later.
+            List<RemainingAssignmentsForMember<String, Integer>> potentiallyUnfilledMembers = new ArrayList<>();
+
+            AssignmentTopicMetadata topicData = assignmentSpec.topics().get(topicId);
+            int numPartitionsForTopic = topicData.numPartitions();
+
+            // Idle members case : When the number of members subscribed to a topic is greater than the total number of Partitions,
+            // all members get assigned via the "extra partitions" logic since minRequiredQuota = 0.
+            int minRequiredQuota = numPartitionsForTopic / membersForTopic.size();
+            // Each member can get only ONE extra partition per topic after receiving the minimum quota.
+            int numMembersWithExtraPartition = numPartitionsForTopic % membersForTopic.size();
+
+            for (String memberId: membersForTopic) {
+                // The partitions need to be in numeric order since we want the same partition numbers from each topic
+                // to go to the same member in case of co-partitioned topics to facilitate joins.
+                Set<Integer> assignedPartitionsForTopic =  assignmentSpec.members().get(memberId).assignedPartitions().getOrDefault(topicId, new HashSet<>());
+
+                int currentAssignmentSize = assignedPartitionsForTopic.size();
+                List<Integer> currentAssignmentListForTopic = new ArrayList<>(assignedPartitionsForTopic);
+
+                // If there were previously assigned partitions present, we want to retain them.
+                if (currentAssignmentSize > 0) {
+                    // We either need to retain currentSize number of partitions when currentSize < required OR required number of partitions otherwise.
+                    int retainedPartitionsCount = min(currentAssignmentSize, minRequiredQuota);
+                    Collections.sort(currentAssignmentListForTopic);
+                    for (int i = 0; i < retainedPartitionsCount; i++) {
+                        assignedStickyPartitionsPerTopic.computeIfAbsent(topicId, k -> new HashSet<>()).add(currentAssignmentListForTopic.get(i));
```

Comment:
nit: A small stylistic comment: It would be better to structure such line as follow: 

```
assignedStickyPartitionsPerTopic
    .computeIfAbsent(topicId, k -> new HashSet<>())
    .add(currentAssignmentListForTopic.get(i));
```

I find this more readable than those long lines.